### PR TITLE
`provider-nop` spring cleaning!

### LINKED ISSUE #5: Support providing arbitrary fields in spec
### What problem are you facing?

A potential use of `provider-nop` is testing behavior of different types of patches in composition. Currently, only fields that set the conditions of the resource are included in the `NopResource` spec, but it could be extended to support arbitrary fields -- effectively allowing it to be a drop in `base` for a set of patches users may want to test on another resource, that potentially would be costly to actually provision (such as an `RDSInstance`).

### How could Crossplane help solve your problem?

Add support for arbitrary fields to the `NopResource` spec.

### PR DESCRIPTION
**Pull request by negz**
### Description of your changes

Fixes https://github.com/crossplane-contrib/provider-nop/issues/5

I'm working on updating Crossplane E2E tests in https://github.com/crossplane/crossplane/pull/4187. I want to keep using `provider-nop` in these tests. To that end I figured I'd update it. The updates include:

* New `spec.forProvider.fields` and `status.atProvider.fields` - arbitrary objects to patch to/from.
* New `spec.forProvider.connectionDetails` - arbitrary connection details to emit.
* Use the `managed.Reconciler` to more closely emulate the behaviour a real provider.
* Build a modern layered `xpkg`, push it to xpkg.upbound.io/crossplane-contrib and index.docker.io/crossplanecontrib

Note this implies two important behaviour changes:

* The managed reconciler will manage the `Synced` condition separately from `conditionAfter`.
* We'll no longer push images to index.docker.io/crossplane.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

Deployed the file from `examples/` to a kind cluster and verified that:

* `spec.forProvider.fields` are accepted.
* `spec.forProvider.connectionDetails` are written to the secret
* `spec.forProvider.conditionAfter` conditions are created

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -292,24 +297,32 @@ jobs:
...
           GIT_API_TOKEN: ${{ secrets.GITHUB_TOKEN }}
 
       - name: Promote Artifacts in S3 and Docker Hub
-        if: github.ref == 'refs/heads/main' && env.AWS_USR != '' && env.DOCKER_USR != ''
+        if: github.ref == 'refs/heads/main' && env.AWS_USR != '' && env.CONTRIB_DOCKER_USR != ''
         run: make -j2 promote
         env:
           BRANCH_NAME: main
-          CHANNEL: main
+          CHANNEL: master
```
Comment (hasheddan) [inline .github/workflows/ci.yaml]: Is this change intentional / required?

**Review by hasheddan — COMMENTED** [review]
Thanks @negz! One question below 👍🏻

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (negz) [reply .github/workflows/ci.yaml]: No, this is a product of cargo-culting these files from another repo to update them to modern practices. I went through `s/master/main` but must have missed this.

Comment (negz) [reply .github/workflows/ci.yaml]: Fixed

**Review by hasheddan — APPROVED** [review]
Thanks @negz !
